Cgmes active reactive regulating control

nemanja-st opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**Does this PR already have an issue describing the problem?**
Fixes #2920 '

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [x] No

**Other information**:
`ACTIVE_POWER` and `CURRENT_LIMITER` are already implemented for PSTs.

nemanja-st commented:
Issues observed for this PR:
1. StaticVarCompensator:
     - When `StaticVarCompensatorRegulationMode` is `OFF` and `regulatingTerminal` is **remote**, `RegulatingControl` is always exported with `enabled` set to false.
     - When `StaticVarCompensatorRegulationMode` is `OFF` and `regulatingTerminal` is **local**, `RegulatingControl` is never exported.
2. ShuntCompensator:
    - When `voltageRegulatorOn` is `false` and `regulatingTerminal` is **remote**, `RegulatingControl` is always exported with `enabled` set to false.
    - When `voltageRegulatorOn` is `false` and `regulatingTerminal` is **local**, `RegulatingControl` is never exported.

annetill commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/EquipmentExport.java:
```diff
@@ -582,13 +594,35 @@ private static void writeShuntCompensators(Network network, Map<Terminal, String
+    private static String getSvcMode(StaticVarCompensator svc) {
+        String mode = RegulatingControlEq.REGULATING_CONTROL_VOLTAGE;
+        StaticVarCompensator.RegulationMode regulationMode = svc.getRegulationMode();
+        // FIXME: remove RegulationMode.OFF when #2790 is done
```
#2790 will not be available soon because it leads to an increase of iidm version. Maybe in June, but no sure at all. Do you have a unit test with a OFF mode but valid values for both voltage and reactive power control values? What is exported then?

marqueslanauja commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/CgmesExportContext.java:
```diff
@@ -533,7 +533,12 @@ private void addIidmMappingsShuntCompensators(Network network) {
+                    svc.getRegulationMode() == StaticVarCompensator.RegulationMode.VOLTAGE ||
+                            svc.getRegulationMode() == StaticVarCompensator.RegulationMode.REACTIVE_POWER ||
+                            validVoltageSetpoint && !validReactiveSetpoint || !validVoltageSetpoint && validReactiveSetpoint)) {
                 regulatingControlId = namingStrategy.getCgmesId(ref(svc), Part.REGULATING_CONTROL);
```
In concordance with the previous comment, we can export it if:

```
if (regulatingControlId == null && (validVoltageSetpoint
                            || validReactiveSetpoint
                            || !Objects.equals(svc, svc.getRegulatingTerminal().getConnectable()))) {
```

nemanja-st replied:
Again, this is related to RC creation from scratch (no CGMES import).
In CGMES, SVC must have a RC in order to function (SVC parameters regarding control should be ignored). This is valid for local and remote controls.
In this implementation, we avoid exporting RC only when regulationMode is not set and we have both voltage and reactive targets (it's not clear which to use).

annetill replied:
We have a use case where we export only an EQ file, without SSH file.

marqueslanauja commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/EquipmentExport.java:
```diff
@@ -365,7 +367,15 @@ private static void writeGenerators(Network network, Map<Terminal, String> mapTe
+                regulatingTerminal = rrpc.getRegulatingTerminal();
+                mode = RegulatingControlEq.REGULATING_CONTROL_REACTIVE_POWER;
+            }
             switch (cgmesOriginalClass) {
```
My proposal:

```
RemoteReactivePowerControl rrpc = generator.getExtension(RemoteReactivePowerControl.class);
String mode = getGeneratorRegulatingControlMode(generator, rrpc);
Terminal regulatingTerminal = mode.equals(RegulatingControlEq.REGULATING_CONTROL_VOLTAGE) ? generator.getRegulatingTerminal() : rrpc.getRegulatingTerminal();
```

```
private static String getGeneratorRegulatingControlMode(Generator generator, RemoteReactivePowerControl rrpc) {
        if (rrpc == null) {
            return RegulatingControlEq.REGULATING_CONTROL_VOLTAGE;
        }
        boolean enabledVoltageControl = generator.isVoltageRegulatorOn();
        boolean enabledReactivePowerControl = rrpc.isEnabled();
        
        if (enabledVoltageControl) {
            return RegulatingControlEq.REGULATING_CONTROL_VOLTAGE;
        } else if (enabledReactivePowerControl) {
            return RegulatingControlEq.REGULATING_CONTROL_REACTIVE_POWER;
        } else {
            boolean validVoltageSetpoint = SteadyStateHypothesisExport.isValidVoltageSetpoint(generator.getTargetV());
            if (validVoltageSetpoint) {
                return RegulatingControlEq.REGULATING_CONTROL_VOLTAGE;
            } else {
                return RegulatingControlEq.REGULATING_CONTROL_REACTIVE_POWER;
            }
        }
    }
```
We can define the `getGeneratorRegulatingControlMode` method in the `CgmesExportUtil` class and use it during the `SSH` export to be coherent with the `EQ` export.

nemanja-st replied:
The difficulty here is in priorities: Is the RRPC extension priority or voltage control?

Example: rrpc exists, voltage control is enabled -> We export voltage control, regardless of rrpc status.

Since rrpc  has to be created explicitly as an extension, assumption is that it has the priority.
Therefore, if rrpc is there, we ignore voltage control completely and only use rrpc parameters.

annetill replied:
If the reactive power control extension is present and enabled, it does not override the voltage control if enabled. Maybe it is a mistake but it is what is developed inside open loadflow.

marqueslanauja replied:
I prefer to select the control priority using the electrical information rather than the  software information. The final user, in most of the cases, will not know how the model is internally implemented. Then, if both controls have the same valid attributes I prefer to export the voltage control. But as before, both options are valid.

marqueslanauja commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/EquipmentExport.java:
```diff
@@ -582,13 +594,26 @@ private static void writeShuntCompensators(Network network, Map<Terminal, String
+        return mode;
+    }
+
     private static void writeLines(Network network, Map<Terminal, String> mapTerminal2Id, String cimNamespace, String euNamespace, String valueAttributeName, String limitTypeAttributeName, String limitKindClassName, boolean writeInfiniteDuration, XMLStreamWriter writer, CgmesExportContext context) throws XMLStreamException {
```
My proposal:

```
private static String getSvcMode(StaticVarCompensator svc) {
        if (svc.getRegulationMode().equals(StaticVarCompensator.RegulationMode.VOLTAGE)) {
            return RegulatingControlEq.REGULATING_CONTROL_VOLTAGE;
        } else if (svc.getRegulationMode().equals(StaticVarCompensator.RegulationMode.REACTIVE_POWER)) {
            return RegulatingControlEq.REGULATING_CONTROL_REACTIVE_POWER;
        } else {
            if (SteadyStateHypothesisExport.isValidReactivePowerSetpoint(svc.getReactivePowerSetpoint())) {
                return RegulatingControlEq.REGULATING_CONTROL_VOLTAGE;
            } else {
                return RegulatingControlEq.REGULATING_CONTROL_REACTIVE_POWER
            }
        }
    }
```

We can also define the getSvcMode method in the CgmesExportUtil class and use it during the SSH export to be coherent with the EQ export.

nemanja-st replied:
Same comment as above, when SVC control mode is not defined, and we have valid voltage and reactive targets, the control shouldn't be exported.
But I agree it should be a common method in `CgmesExportUtil`.

marqueslanauja replied:
I am in favor of exporting the maximum information and in this case I prefer to export the voltage control as disabled but both options are valid.

marqueslanauja commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/EquipmentExport.java:
```diff
@@ -904,8 +929,8 @@ private static <C extends Connectable<C>> void writeRatioTapChanger(C eq, RatioT
-                    TapChangerEq.writeControl(cgmesRegulatingControlId, controlName, RATIO_TAP_CHANGER_REGULATION_MODE_VOLTAGE, terminalId, cimNamespace, writer, context);
+                    String tccMode = (rtc.getRegulationMode() == RatioTapChanger.RegulationMode.REACTIVE_POWER) ? RATIO_TAP_CHANGER_REGULATION_MODE_REACTIVE_POWER : RATIO_TAP_CHANGER_REGULATION_MODE_VOLTAGE;
+                    TapChangerEq.writeControl(cgmesRegulatingControlId, controlName, tccMode, terminalId, cimNamespace, writer, context);
                     regulatingControlsWritten.add(cgmesRegulatingControlId);
```
We can use a method to obtain the tccMode and define it in the CgmesExportUtil class and use it during the SSH export to be always coherent with the EQ export.

marqueslanauja commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/SteadyStateHypothesisExport.java:
```diff
@@ -184,6 +185,7 @@ private static String cgmesTapChangerId(TwoWindingsTransformer twt, String tapCh
     private static void writeTapChangers(Network network, String cimNamespace, Map<String, List<RegulatingControlView>> regulatingControlViews, XMLStreamWriter writer, CgmesExportContext context) throws XMLStreamException {
         for (TwoWindingsTransformer twt : network.getTwoWindingsTransformers()) {
+            CgmesExportUtil.addUpdateCgmesTapChangerExtension(twt, context);
             if (twt.hasPhaseTapChanger()) {
```
I suppose that you added to be able to export the SSH file without exporting the EQ file.
You also have to add this method some lines below for the threeWindingsTransformers.

nemanja-st replied:
Thanks, I'll update that!

marqueslanauja commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/SteadyStateHypothesisExport.java:
```diff
@@ -365,18 +367,33 @@ private static void addRegulatingControlView(Generator g, Map<String, List<Regul
+                targetValueUnitMultiplier = "k";
+                enabled = g.isVoltageRegulatorOn();
+            }
+
```
Use the `getGeneratorRegulatingControlMode` method to obtain the mode, then depending on the mode set the attributes

marqueslanauja commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/SteadyStateHypothesisExport.java:
```diff
@@ -365,18 +367,33 @@ private static void addRegulatingControlView(Generator g, Map<String, List<Regul
+    public static boolean isValidVoltageSetpoint(double v) {
         return Double.isFinite(v) && v > 0;
     }
 
```
Maybe move this method to the CgmesExportUtil class given that is used in several classes

marqueslanauja commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/SteadyStateHypothesisExport.java:
```diff
@@ -365,18 +367,33 @@ private static void addRegulatingControlView(Generator g, Map<String, List<Regul
+    public static boolean isValidReactivePowerSetpoint(double q) {
         return Double.isFinite(q);
     }
 
```
Maybe move also to the CgmesExportUtil class

marqueslanauja commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/SteadyStateHypothesisExport.java:
```diff
@@ -402,11 +419,11 @@ private static void writeStaticVarCompensators(Network network, String cimNamesp
                 } else if (regulationMode == StaticVarCompensator.RegulationMode.REACTIVE_POWER
-                        || regulationMode == StaticVarCompensator.RegulationMode.OFF && isValidSvcReactivePowerSetpoint(svc.getReactivePowerSetpoint())) {
+                        || regulationMode == StaticVarCompensator.RegulationMode.OFF && isValidReactivePowerSetpoint(svc.getReactivePowerSetpoint()) && !isValidVoltageSetpoint(svc.getVoltageSetpoint())) {
                     targetValue = svc.getReactivePowerSetpoint();
```
Maybe use the getSvcMode to obtain the mode, then depending on the mode set the attributes

marqueslanauja commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/SteadyStateHypothesisExport.java:
```diff
@@ -443,14 +460,17 @@ private static void addRegulatingControlView(TapChanger<?, ?, ?, ?> tc, CgmesTap
+                    case VOLTAGE -> "k";
+                    case REACTIVE_POWER -> "M";
+                };
                 rcv = new RegulatingControlView(controlId,
```
Maybe use the method used in the EQ export to obtain the mode and then depending on the mode set the unitMultiplier

nemanja-st commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/CgmesExportContext.java:
```diff
@@ -523,7 +523,7 @@ private void addIidmMappingsShuntCompensators(Network network) {
             String regulatingControlId = shuntCompensator.getProperty(Conversion.CGMES_PREFIX_ALIAS_PROPERTIES + REGULATING_CONTROL);
-            if (regulatingControlId == null && (shuntCompensator.isVoltageRegulatorOn() || !Objects.equals(shuntCompensator, shuntCompensator.getRegulatingTerminal().getConnectable()))) {
+            if (regulatingControlId == null && SteadyStateHypothesisExport.isValidVoltageSetpoint(shuntCompensator.getTargetV())) {
                 regulatingControlId = namingStrategy.getCgmesId(ref(shuntCompensator), Part.REGULATING_CONTROL);
```
This is RC creation from scratch (`regulatingControlId == null`) so it comes down to deciding if Shunt has voltage regulating capability from IIDM.
If targetV is not set, in case of full CGMES export from IIDM, SSH will have incorrect values for the control.

marqueslanauja replied:
If the `targetV` is not set, the validation level is EQUIPMENT and does not make sense to export the SSH file. 
I prefer to export properly the EQ file, including the control, in concordance with the validation level.

nemanja-st commented:
Hello @marqueslanauja @annetill 
I implemented the suggestions, refactored the code and fixed some tests.
The behavior on control export currently:
- Generator:
    - Imported control is always exported.
    - No control when no reactive capability (Qmin=Qmax)
    - Reactive control when reactive is enabled and voltage disabled
    - Voltage in all other cases
- Shunt:
    - No control when control is not remote and no voltage setpoint
    - Voltage exported in all other cases
- SVC:
    - No control when control is local, OFF and there are no valid Q or V targets
    - Reactive control when control is REACTIVE or control is OFF and reactive power target is valid and voltage target is invalid
    - Voltage control in all other cases
- RTC:
    - Control is not exported when it's not defined (RatioRegulationMode not defined), **regardless if it has remote terminal or not**.
    - Otherwise, control exported based on RatioRegulationMode

annetill commented:
Hi @nemanja-st, documentation is missing
